### PR TITLE
Introduce ability to use grouping for `leftJoin`

### PR DESCRIPTION
**Pull request by ttim**
This PR adds `LeftJoinGrouping` options which makes flat map nodes which correspond to `LeftJoinProducer` grouped by key. This achieved by couple of changes:
1. introduction of `LeftJoinGrouping` option itself
2. support this option in `OnlinePlan` to force separate node for grouped `leftJoin` producers
3. support grouped `Item[(K, V)]` tuples in `StormTopologyBuilder`
4. tests for this in `StormLaws` and `TopologyTests`.

### CODE REVIEW

**Comment by ttim** [comment]
Also I tested this change internally at twitter and looks like it works.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala [hunk 1]
```diff
@@ -220,6 +221,23 @@ class StormLaws extends WordSpec {
     }
   }
 
+  "StormPlatform should work with grouped leftJoin" in {
+    val leftJoinName = "leftJoin"
+    val producer = TestPlatform.source(sample[List[Int]])
+      .map(sample[(Int) => (Int, Int)])
```
Comment (pankajroark) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: Parens around the first Int look redundant.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/TopologyTests.scala [hunk 2]
```diff
@@ -312,4 +315,39 @@ class TopologyTests extends WordSpec {
...
+
+  "Grouped leftJoin creates fields grouping" in {
+    val leftJoinName = "leftJoin"
+    val producer: TailProducer[Storm, _] =
+      Storm.source(TraversableSpout(sample[List[Int]]))
+        .flatMap(testFn)
+        .leftJoin(
+          ReadableServiceFactory[Int, Int](() => ReadableStore.fromFn(v => Some(v)))
+        ).name(leftJoinName)
+        .map(v => (v._1, v._2._1))
```
Comment (pankajroark) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/TopologyTests.scala]: Hard to read, may be use deconstruction. map { case (k, v) => ...

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala [hunk 3]
```diff
@@ -113,4 +139,19 @@ private object EdgeInjections {
       (input.get(0).asInstanceOf[T1], input.get(1).asInstanceOf[T2], input.get(2).asInstanceOf[T3])
     }
   }
+
+  case class KeyValueInjection[K, V]() extends Injection[Item[(K, V)], JList[AnyRef]] {
+    override def apply(tuple: Item[(K, V)]): JList[AnyRef] = {
+      val list = new JAList[AnyRef](3)
+      list.add(tuple._1.asInstanceOf[AnyRef])
```
Comment (pankajroark) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala]: Destructure for readability.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala [hunk 4]
```diff
@@ -113,4 +139,19 @@ private object EdgeInjections {
...
+      list.add(tuple._2._1.asInstanceOf[AnyRef])
+      list.add(tuple._2._2.asInstanceOf[AnyRef])
+      list
+    }
+
+    override def invert(valueIn: JList[AnyRef]): Try[Item[(K, V)]] = Inversion.attempt(valueIn) { input =>
+      (input.get(0).asInstanceOf[Timestamp],
```
Comment (pankajroark) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala]: May be create temporary variables for readability and then
(ts, (k, v))
which is much more readable.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala [hunk 5]
```diff
@@ -179,10 +173,21 @@ abstract class Storm(options: Map[String, Options], transformConfig: Summingbird
       }
     }
 
-    PlannedTopology(config, StormTopologyBuilder(options, jobId, stormDag).build)
+    val topology = StormTopologyBuilder(options, jobId, stormDag).build
```
Comment (pankajroark) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala:176]: Nice, this reads simpler.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala [hunk 6]
```diff
@@ -179,10 +173,21 @@ abstract class Storm(options: Map[String, Options], transformConfig: Summingbird
...
+    PlannedTopology(config, topology.build(jobId))
   }
   def run(tail: TailProducer[Storm, _], jobName: String): Unit = run(plan(tail), jobName)
   def run(plannedTopology: PlannedTopology, jobName: String): Unit
+
+  private[storm] def planDag[T](tail: TailProducer[Storm, T]): Dag[Storm] = {
+    /*
+     * TODO: storm does not yet know about ValueFlatMapped, so remove it before
```
Comment (pankajroark) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala]: Can you create a jira ticket and link here?

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala [hunk 7]
```diff
@@ -144,8 +142,35 @@ private[storm] case class StormTopologyBuilder(options: Map[String, Options], jo
     case Some(props) =>
       registerShardedKeyValue[K, V](topology, node, provider, props.batcher, props.shards)
     case None =>
-      throw new IllegalStateException("Shouldn't happen, trying to register component which emits " +
-        "keyed values and don't have downstream summers.")
+      registerKeyValue[K, V](topology, node, provider)
```
Comment (pankajroark) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:145]: nit: I just noticed this method is becoming hard to read due to nested deconstructors, perhaps refactor "case Some(props) if props.allSummers =>" part into a separate method.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala [hunk 8]
```diff
@@ -216,14 +243,23 @@ private[storm] case class StormTopologyBuilder(options: Map[String, Options], jo
...
+      case summerNode: SummerNode[Storm] => true
+    }
 
   private def withLocalGrouping(node: FlatMapNode[Storm]): Boolean = {
     val usePreferLocalDependency = getOrElse(node, DEFAULT_FM_PREFER_LOCAL_DEPENDENCY)
     logger.info(s"[${getNodeName(node)}] usePreferLocalDependency: ${usePreferLocalDependency.get}")
     usePreferLocalDependency.get
   }
 
+  private def isGroupedLeftJoinNode(node: FlatMapNode[Storm]): Boolean =
```
Comment (pankajroark) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:257]: This doesn't need to be a method on this class. Should perhaps be a method on the companion and unit tested.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala [hunk 9]
```diff
@@ -92,16 +93,13 @@ private[storm] object StormTopologyBuilder {
...
-    val topology = stormDag.nodes.foldLeft(Topology.empty) {
-      case (currentTopology, node: SummerNode[Storm]) =>
-        register(currentTopology, node, SummerBoltProvider(this, node))
-      case (currentTopology, node: FlatMapNode[Storm]) =>
-        register(currentTopology, node, FlatMapBoltProvider(this, node))
-      case (currentTopology, node: SourceNode[Storm]) =>
-        register(currentTopology, node, SpoutProvider(this, node))
-    }
-    topology.build(jobId)
+  def build: Topology = stormDag.nodes.foldLeft(Topology.empty) {
```
Comment (pankajroark) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:96]: We should add unit tests for StormTopologyBuilder to cover this feature.

**Review by pankajroark — CHANGES_REQUESTED** [review]
Looks great overall. I think good to ship after adding unit tests for StormTopologyBuilder.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala (hunk 1 above)
Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: Done

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala (hunk 3 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala]: Done

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/TopologyTests.scala (hunk 2 above)
Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/TopologyTests.scala]: Done

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala (hunk 4 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala]: Done

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala (hunk 6 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala]: Actually this is not needed anymore, removed this code and added test for `mapValues` into `StormLaws`. Thanks!

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 9 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:96]: I added test to `TopologyTests`, but I think separate `StormTopologyBuilderTests` would be nicer.

Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:96]: Done

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 8 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:257]: Unfortunately in this method we access `Options` which makes both extraction and testing much harder. I'm agree this is not the best place for this method but IMO it's best place now.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 7 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:145]: This method represents high level strategy how to deal with key value components, all further logic extracted into `registerAggregatedEdges`, `registerShardedKeyValue` and `registerKeyValue`. I can extract `aggregated` part further but I prefer to keep it as is. Also I doubt it will grow more.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala [hunk 10]
```diff
@@ -162,13 +162,7 @@ abstract class Storm(options: Map[String, Options], transformConfig: Summingbird
   def withConfigUpdater(fn: SummingbirdConfig => SummingbirdConfig): Storm
 
   def plan[T](tail: TailProducer[Storm, T]): PlannedTopology = {
-    /*
-     * TODO: storm does not yet know about ValueFlatMapped, so remove it before
-     * planning
-     */
-    val dagOptimizer = new DagOptimizer[Storm] {}
-    val stormTail = dagOptimizer.optimize(tail, dagOptimizer.ValueFlatMapToFlatMap)
```
Comment (johnynek) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala:170]: this was already done? We have support for ValueFlatMapped now? If so, can we combine `leftJoin.mapValues.sumByKey` into a single node if `leftJoin` is grouped?

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala [hunk 11]
```diff
@@ -216,14 +243,23 @@ private[storm] case class StormTopologyBuilder(options: Map[String, Options], jo
...
+    }
 
   private def withLocalGrouping(node: FlatMapNode[Storm]): Boolean = {
     val usePreferLocalDependency = getOrElse(node, DEFAULT_FM_PREFER_LOCAL_DEPENDENCY)
     logger.info(s"[${getNodeName(node)}] usePreferLocalDependency: ${usePreferLocalDependency.get}")
     usePreferLocalDependency.get
   }
 
+  private def isGroupedLeftJoinNode(node: FlatMapNode[Storm]): Boolean =
+    node.members.last.isInstanceOf[LeftJoinedProducer[Storm, _, _, _]] &&
```
Comment (johnynek) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala]: I'm suspect of the `.last` here. This means we can't add a `mapValues` or `flatMapValues` on the end.

Why don't we want `node.members.exists { case _: LeftJoinedProducer[_, _, _, _] => true; case _ => false } && `?

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTopologyBuilderTests.scala [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+  "Grouped leftJoin creates fields grouping" in {
+    val leftJoinName = "leftJoin"
+    val producer: TailProducer[Storm, _] =
+      Storm.source(TraversableSpout(sample[List[Int]]))
+        .flatMap(testFn)
+        .leftJoin(
+          ReadableServiceFactory[Int, Int](() => ReadableStore.fromFn(v => Some(v)))
+        ).name(leftJoinName)
+        .map { case (key, (value, optPrevValue)) => (key, value) }
```
Comment (johnynek) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTopologyBuilderTests.scala]: can you write this with `.mapValues { case (value, _) => value }`, which has the potential for optimization in scalding at least, so people can see examples where `mapValues/collectValues/flatMapValues` is preferred?

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala [hunk 13]
```diff
@@ -113,4 +139,20 @@ private object EdgeInjections {
...
+      case (timestamp, (key, value)) =>
+        val list = new JAList[AnyRef](3)
+        list.add(timestamp.asInstanceOf[AnyRef])
+        list.add(key.asInstanceOf[AnyRef])
+        list.add(value.asInstanceOf[AnyRef])
+        list
+    }
+
+    override def invert(valueIn: JList[AnyRef]): Try[Item[(K, V)]] = Inversion.attempt(valueIn) { input =>
+      val (timestamp, key, value) = (input.get(0), input.get(1), input.get(2))
```
Comment (pankajroark) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala]: I'm not sure scalac will optimize this out, so there could be performance penalty. May be just split to multiple lines. 
val timestamp = input.get(0).asInstanceOf[Timestamp]
...

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 7 above)
Comment (pankajroark) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:145]: sg

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 8 above)
Comment (pankajroark) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala:257]: sg, since extraction is harder it's ok.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 11 above)
Comment (pankajroark) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala]: It would also be good to add tests with mapValues and flatMapValues to capture the edge cases that @johnynek pointed out.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala (hunk 10 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala:170]: We have support on the level we can build `flatMap` operation over it (https://github.com/twitter/summingbird/commit/44fb5c9#diff-1d4313fb460f1921fc349a8eb2ee902fR85). But it's not fully supported in a way to affect `OnlinePlan`. I would like to do this in future, but I guess it's out of this PR. It's about when do not trigger new node creation, and this one is when to trigger it =)

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 11 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala]: @johnynek `.last` is here refers to first element in producers chain unfortunately (I added a comment on `.members` in previous PR).

I don't think we need tests on this edge cases because we don't support this edge cases yet.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTopologyBuilderTests.scala (hunk 12 above)
Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTopologyBuilderTests.scala]: Sure. Looks better now, thanks!

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala (hunk 13 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala]: I thought about this but decided it should be optimized by compiler =) changed to your way.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 11 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala]: Added a comment in test that number of bolts can be `-1` in case of support of `mapValues`.

Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala]: Also I added `firstProducer` method on `Dag`s `Node`, to make it look better.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala (hunk 10 above)
Comment (pankajroark) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala:170]: I agree, better to go incrementally. We can do that optimization in separate review.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala [hunk 14]
```diff
@@ -307,7 +307,7 @@ private[storm] case class StormTopologyBuilder(options: Map[String, Options], jo
     }
 
   private[storm] def get[T <: AnyRef: ClassTag](node: StormNode): Option[(String, T)] = {
-    val producer = node.members.last
+    val producer = node.firstProducer.get
```
Comment (johnynek) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala]: any reason not to do:
```scala
node.firstProducer.flatMap { producer =>
  Options.getFirst[T](options, stormDag.producerToPriorityNames(producer))
}
```
here?

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala [hunk 15]
```diff
@@ -48,6 +60,17 @@ private[storm] object Edges {
     destId
   )
 
+  def groupedShardedToItem[K, V](
+    sourceId: Topology.EmittingId[Sharded[K, V]],
+    destId: Topology.ReceivingId[Item[(K, V)]]
+  ): Topology.Edge[Sharded[K, V], Item[(K, V)]] = Topology.Edge(
```
Comment (johnynek) [inline summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala:66]: this function is untested. Any way to get some test coverage of it?

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala (hunk 10 above)
Comment (johnynek) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala:170]: wait, I'm confused why you removed this code then.

Can you link to where in the storm platform we handle `flatMapValues`? The link you have above points to the optimizer rule you are removing here.

**Review by pankajroark — APPROVED** [review]
LGTM

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala (hunk 10 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormPlatform.scala:170]: @johnynek because I pointed to the wrong code, sorry %) Here is a right link https://github.com/twitter/summingbird/commit/44fb5c9#diff-81c566a323153de58aa7f47b84ee941aR54.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala (hunk 14 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala]: Changed to `.flatMap` way.

File: summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala (hunk 15 above)
Comment (ttim) [reply summingbird-storm/src/main/scala/com/twitter/summingbird/storm/Edges.scala:66]: Sure, done!

**Comment by johnynek** [comment]
👍 

100% test coverage of this diff. Nice!

**Comment by johnynek** [comment]
It will be exciting when we can merge `leftJoin.mapValues.sumByKey` into one node to reduce serialization.

I bet this is actually a pretty common pattern: look-up by the key, do some transformation, aggregate a value.

**Comment by pankajroark** [comment]
Agree with @johnynek most of our realtime topologies have sumByKey and leftJoin when present is almost always a preparatory step for it, so leftJoin.mapValues.sumByKey is pretty common when leftJoin is used. That will be a very useful optimization.
